pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -252,7 +253,38 @@ public void uploadSegmentAsMultiPart(FormDataMultiPart multiPart,
       @ApiParam(value = "Whether to enable parallel push protection") @DefaultValue("false") @QueryParam(FileUploadDownloadClient.QueryParameters.ENABLE_PARALLEL_PUSH_PROTECTION) boolean enableParallelPushProtection,
       @Context HttpHeaders headers, @Context Request request, @Suspended final AsyncResponse asyncResponse) {
     try {
-      asyncResponse.resume(uploadSegmentInternal(multiPart, null, enableParallelPushProtection, headers, request));
+      asyncResponse.resume(uploadSegmentInternal(multiPart, null, enableParallelPushProtection, headers, request, false));
+    } catch (Throwable t) {
+      asyncResponse.resume(t);
+    }
+  }
+
+  /**
+   * Will upload a Pinot segment without sending the file through the controller. If data is stored on a separate
+   * filesystem, eg: HDFS or Azure, and the segment file is already in its final location, users can call this method
+   * to change the state in zk.
+   * @param multiPart
+   * @param enableParallelPushProtection
+   * @param headers
+   * @param request
+   * @param asyncResponse
+   */
+  @POST
+  @ManagedAsync
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.MULTIPART_FORM_DATA)
+  @Path("/segmentmetadata")
+  @ApiOperation(value = "Upload a segment by sending metadata", notes = "Does not download segment file")
+  public void uploadSegmentAsMultiPartWithMetadata(FormDataMultiPart multiPart,
+      @ApiParam(value = "Whether to enable parallel push protection") @DefaultValue("false") @QueryParam(FileUploadDownloadClient.QueryParameters.ENABLE_PARALLEL_PUSH_PROTECTION) boolean enableParallelPushProtection,
```

Comment:
Why is enabling parallel push protection a param for POST? What happens if multiple posts are called concurrently and some of them enable this param and others disable it? Would it make more sense to have that as a controller/table config instead of per POST request?

Reply:
We've done this for the tar segment file upload as well as the JSON file upload. The idea is that we don't do the extra zk write (that would protect against mutant metadata) unless clients send that header to protect against zk problems. We can change this in a future pr, but for now, this is consistent with other segment uploads